README.md
# Edges Eraser

[![PyPI](https://img.shields.io/pypi/v/edgeseraser?style=flat-square)](https://pypi.python.org/pypi/edgeseraser/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/edgeseraser?style=flat-square)](https://pypi.python.org/pypi/edgeseraser/)
[![PyPI - License](https://img.shields.io/pypi/l/edgeseraser?style=flat-square)](https://pypi.python.org/pypi/edgeseraser/)

[![Coookiecutter - Wolt](https://img.shields.io/badge/cookiecutter-Wolt-00c2e8?style=flat-square&logo=cookiecutter&logoColor=D4AA00&link=https://github.com/woltapp/wolt-python-package-cookiecutter)](https://github.com/woltapp/wolt-python-package-cookiecutter)


---

**Documentation**: [https://devmessias.github.io/edgeseraser](https://devmessias.github.io/edgeseraser)

**Source Code**: [https://github.com/devmessias/edgeseraser](https://github.com/devmessias/edgeseraser)

**PyPI**: [https://pypi.org/project/edgeseraser/](https://pypi.org/project/edgeseraser/)

---

## What is Edges Eraser?
This pkg aims to implement serveral filtering methods for (un)directed graphs.

Edge filtering methods allows to extract the backbone of a graph or sampling the most important edges. You can use edge filtering methods as a preprocessing step aiming to improve the performance/results of graph algorithms or to turn a graph visualtzation more asthetic.


## Example
```python
import networkx as nx
import edgeseraser as ee

g = nx.erdos_renyi_graph(100, 0.1)
ee.noise_score.filter_nx_graph(g)

g # filtered graph
```

## Available methods and details

| Method | Description | suitable for | limitations/restrictions/details |
| --- | --- |--- | --- |
| [Noise Score] | Filters edges with high noise score. Paper:[1]|Directed, Undirected, Weighted | Very good and fast! [4] |
| [Disparity] | Dirichlet process filter (stick-breaking) Paper:[2] |  Directed, Undirected, Weighted |There are some criticism regarding the use in undirected graphs[3]|
| [Pólya-Urn]| Filters edges with Pólya-Urn method. Paper:[5]| Directed, Undirected, Integer Weighted||

[1]: https://arxiv.org/abs/1701.07336
[2]: https://arxiv.org/abs/0904.
[3]: https://arxiv.org/abs/2101.00863
[4]: https://www.michelecoscia.com/?p=1236
[5]: https://www.nature.com/articles/s41467-019-08667-3
[Noise Score]: /api_docs/#edgeseraser.noise_score
[Disparity]: /api_docs/#edgeseraser.disparity


## Installation

```sh
pip install edgeseraser
```

## Development

* Clone/Fork this repository

```sh
git clone https://github.com/devmessias/edgeseraser
```

* Requirements:
  * [Poetry](https://python-poetry.org/)
  * Python 3.7+

```sh
make install
make init
```

### Testing

```sh
make test
```

To run the static analysis, use the following command:
```sh
make mypy
```

### Documentation

The documentation is automatically generated from the content of the [docs directory](./docs) and from the docstrings
 of the public signatures of the source code. The documentation is updated and published as a [Github project page
 ](https://pages.github.com/) automatically as part each release.

To see the current state of the documentation in your browser, use the following command:
```sh
make docs-serve
```
The above command will start a local server on port 8000. Any changes to
the documentation and docstrings will be automatically reflected in your browser.

### Pre-commit

Pre-commit hooks run all the auto-formatters (e.g. `black`, `isort`), linters (e.g. `mypy`, `flake8`), and other quality
 checks to make sure the changeset is in good shape before a commit/push happens.


If you want e.g. want to run all checks manually for all files:

```sh
make pre-commit
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]

## [0.6.0] - 2022-03-25
### Added
- Static type analysis
  - stubs for networkx, igraph and scipy csr_matrix
  - types for Pólya-Urn method
- Automatic benchmarking with `pytest-benchmark`
- Numba support for special functions GammaLn and BetaLn
- Python's `lru_cache` for memoization inside of Pólya-Urn
- NumbaPolya cache for memoization using JIT-classes
  - This gave a significant speedup for the Pólya-Urn method, approx 80x compared to   numba python-lru version.

### Fixed
- Pólya-Urn performance issues
- Pólya-Urn docstrings

### Changed
- Args for Pólya-Urn method
- Makefile targets:
    -   make tests -> make pytests
        -   skip all the benchmarks tests
    -   make benchmark name="{BENCHMARK_NAME}"
        -   run the benchmarks
    - github actions now uses the makefile for tests
- Return types for all graph filter methods
- Pólya-Urn code organization
- Disable mypy for while until it is fixed

## [0.5.0] - 2022-03-15
### Added
- Pólya-Urn method for integer weighted graphs now uses JIT compilation to
    speed up the computation.
- Tests for integer weighted graphs.
- Makefile targets to build and install the package in a single command.
- Fast math implementation for Beta-log and Gamma-log functions.
- Tests for for fast math implementations.

### Fixed
- Disparity filter
- Pólya-Urn methdo for integer weighted graphs
- Tests

### Changed
- Same variable name for weight, weighted degree, etc.
- Folder structure.

## [0.4.1] - 2022-03-15
### Fixed
- Docstrings
- Missing args inside of `edgeseraser.polya`

## [0.4.0] - 2022-03-14
### Added
- Support for named vertex labels
- Pólya-Urn backbone filter
- Check for disconnected graphs

### Changed
- Organized the edges and weight extraction code using DRY principles

## [0.3.0] - 2022-03-13
### Added
- Python iGraph lib support

### Fixed
- Google docstring return style
- Typos in docstrings

## [0.2.1] - 2022-03-12
### Changed
- Google docstring style

## [0.2.0] - 2022-03-12
### Fixed
- Type hiting issue for python 3.7 using `Literal`

## [0.1.0] - 2022-03-12
### Added
- Noise score filter
- Disparity filter
- Networkx integration

[Unreleased]: https://github.com/devmessias/edgeseraser/compare/0.6.0...master
[0.6.0]: https://github.com/devmessias/edgeseraser/compare/0.5.0...0.6.0
[0.5.0]: https://github.com/devmessias/edgeseraser/compare/0.4.1...0.5.0
[0.4.1]: https://github.com/devmessias/edgeseraser/compare/0.4.0...0.4.1
[0.4.0]: https://github.com/devmessias/edgeseraser/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/devmessias/edgeseraser/compare/0.2.1...0.3.0
[0.2.1]: https://github.com/devmessias/edgeseraser/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/devmessias/edgeseraser/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/devmessias/edgeserase/releases/tag/v0.1.0


edgeseraser/__init__.py
"""This pkg aims to implement serveral filtering methods for (un)directed
graphs.

Edge filtering methods allows to extract the backbone of a graph or sampling the
most important edges. You can use edge filtering methods as a preprocessing step
aiming to improve the performance/results of graph algorithms or to turn a graph
visualtzation more asthetic.

**See the example below for a simple usage of the package.**
```python
import networkx as nx
import edgeseraser as ee

g = nx.erdos_renyi_graph(100, 0.4)
ee.noise_score.filter_nx_graph(g)

g # filtered graph
```

## Available methods and details

| Method | Description | suitable for | limitations/restrictions/details |
| --- | --- |--- | --- |
| [Noise Score] | Filters edges with high noise score. Paper:[1]|Directed, Undirected, Weighted | Very good and fast! [4] |
| [Disparity] | Dirichlet process filter (stick-breaking) Paper:[2] |  Directed, Undirected, Weighted |There are some criticism regarding the use in undirected graphs[3]|
| [Pólya-Urn]| Filters edges with Pólya-Urn method. Paper:[5]| Directed, Undirected, Integer Weighted||

[1]: https://arxiv.org/abs/1701.07336
[2]: https://arxiv.org/abs/0904.
[3]: https://arxiv.org/abs/2101.00863
[4]: https://www.michelecoscia.com/?p=1236
[5]: https://www.nature.com/articles/s41467-019-08667-3

[Noise Score]: https://devmessias.github.io/edgeseraser/api_docs/#edgeseraser.noise_score
[Disparity]: https://devmessias.github.io/edgeseraser/api_docs/#edgeseraser.disparity
[Pólya-Urn]: https://devmessias.github.io/edgeseraser/api_docs/#edgeseraser.polya
"""
__author__ = """Bruno Messias"""
__email__ = "devmessias@gmail.com"
__version__ = "0.5.0"
